res/controllers/Korg-KAOSS-DJ-scripts.js

```diff
@@ -137,68 +168,158 @@ KAOSSDJ.wheelTurnShift = function(channel, control, value, status, group) {
-    var decks = KAOSSDJ.decks;
-    for(key in decks) {
-        var deck = decks[key];
-        if(deck.fx) {
-            engine.setValue('[EffectRack1_EffectUnit'+deck.deckNumber +']', 'super1', value / 127);
+KAOSSDJ.shiftLeftCallback = function(_channel, _control, value, _status, _group) {
+    KAOSSDJ.shiftLeftPressed = value === MIDI_ON;
+};
+
+KAOSSDJ.shiftRightCallback = function(_channel, _control, value, _status, _group) {
+    KAOSSDJ.shiftRightPressed = value === MIDI_ON;
+};
+
+KAOSSDJ.changeFocus = function(_channel, _control, value, _status, _group) {
+    // toggle focus between Playlist and File-Browser
+    if (value === MIDI_ON) {
+        engine.setValue("[Library]", "MoveFocusForward", true);
+    }
+};
+
+// <browseKnob>           : browse library up & down
+// <SHIFT> + <browseKnob> : toggle focus between Playlist and File-Browser
+KAOSSDJ.browseKnob = function(_channel, _control, value, _status, _group) {
+    if (value > 0x40) {
+        if (KAOSSDJ.shiftLeftPressed || KAOSSDJ.shiftRightPressed) {
+            engine.setValue("[Library]", "MoveFocusForward", true);
+        } else {
+            engine.setValue("[Library]", "MoveUp", true);
+        }
+    } else {
+        if (KAOSSDJ.shiftLeftPressed || KAOSSDJ.shiftRightPressed) {
+            engine.setValue("[Library]", "MoveFocusBackward", true);
+        } else {
+            engine.setValue("[Library]", "MoveDown", true);
         }
     }
 };
 
-KAOSSDJ.controllerFxTouchUp = function(channel, control, value, status, group) {
-    var deck = KAOSSDJ.getDeckByChannel(channel);
-    engine.setValue('[EffectRack1_EffectUnit'+deck.deckNumber +']', 'mix', 0);
-    engine.setValue('[EffectRack1_EffectUnit'+deck.deckNumber +']', 'super1', 0);
+// <TAP>                : open folder
+// <TAP> + <TAP>        : double-tap to close folder
+// <SHIFT LEFT> + <TAP> : tap bpm of LEFT track
+// <SHIFT RIGHT> + <TAP> : tap bpm of RIGHT track
+KAOSSDJ.tapButtonCallback = function(_channel, _control, value, _status, _group) {
+    if (value !== MIDI_ON) {
+        return;
+    }
+
+    /* shift tab to move focus view */
+    if ((value === MIDI_ON) && KAOSSDJ.shiftLeftPressed) {
+        // engine.setValue("[Library]", "MoveFocusForward", true);
+        engine.setValue("[Channel1]", "bpm_tap", true);
+        return;
+    }
+    if ((value === MIDI_ON) && KAOSSDJ.shiftRightPressed) {
+        // engine.setValue("[Library]", "MoveFocusForward", true);
+        engine.setValue("[Channel2]", "bpm_tap", true);
+        return;
+    }
+
+    /* tap to open folder, double-tap to close folder (twice to undo first tap)*/
+    const now = new Date();
+    const timeSinceLastTap = now - KAOSSDJ.lastTapDate;
+    KAOSSDJ.lastTapDate = now;
+    if ((timeSinceLastTap < 600) && (timeSinceLastTap > 0)) {
+        engine.setValue("[Library]", "MoveLeft", true);
+    } else {
+        engine.setValue("[Library]", "MoveRight", true);
+    }
```

Comment:
maybe GoToItem + Back would be better?

Reply:
We have no "Back" AFAIK. I leave this for one who can test this.

Reply:
Not really, but we have `[Library], MoveFocusBackward` which is commonly mapped I think.